File: archive/old-modular/serial.c
#include "../include/serial.h"
#include "../include/types.h"

// Port I/O helpers
static inline void outb(uint16_t port, uint8_t val) {
    __asm__ volatile("outb %0, %1" : : "a"(val), "Nd"(port));
}

static inline uint8_t inb(uint16_t port) {
    uint8_t ret;
    __asm__ volatile("inb %1, %0" : "=a"(ret) : "Nd"(port));
    return ret;
}

void serial_init(void) {
    outb(COM1 + 1, 0x00);  // Disable interrupts
    outb(COM1 + 3, 0x80);  // Enable DLAB
    outb(COM1 + 0, 0x03);  // Divisor low byte (38400 baud)
    outb(COM1 + 1, 0x00);  // Divisor high byte
    outb(COM1 + 3, 0x03);  // 8 bits, no parity, one stop bit
    outb(COM1 + 2, 0xC7);  // Enable FIFO, clear, 14-byte threshold
    outb(COM1 + 4, 0x0B);  // IRQs enabled, RTS/DSR set
}

void serial_putc(char c) {
    while ((inb(COM1 + 5) & 0x20) == 0);
    outb(COM1, c);
}

void serial_puts(const char *s) {
    while (*s) {
        if (*s == '\n') serial_putc('\r');
        serial_putc(*s++);
    }
}

// Simple printf implementation
static void print_num(uint64_t num, int base) {
    char buf[32];
    int i = 0;

    if (num == 0) {
        serial_putc('0');
        return;
    }

    while (num > 0) {
        int digit = num % base;
        buf[i++] = (digit < 10) ? ('0' + digit) : ('a' + digit - 10);
        num /= base;
    }

    while (i > 0) {
        serial_putc(buf[--i]);
    }
}

void serial_printf(const char *fmt, ...) {
    __builtin_va_list args;
    __builtin_va_start(args, fmt);

    while (*fmt) {
        if (*fmt == '%') {
            fmt++;
            switch (*fmt) {
                case 'd': {
                    int val = __builtin_va_arg(args, int);
                    if (val < 0) {
                        serial_putc('-');
                        val = -val;
                    }
                    print_num(val, 10);
                    break;
                }
                case 'u': {
                    unsigned int val = __builtin_va_arg(args, unsigned int);
                    print_num(val, 10);
                    break;
                }
                case 'x': {
                    unsigned int val = __builtin_va_arg(args, unsigned int);
                    print_num(val, 16);
                    break;
                }
                case 'l': {
                    fmt++;
                    if (*fmt == 'u') {
                        uint64_t val = __builtin_va_arg(args, uint64_t);
                        print_num(val, 10);
                    } else if (*fmt == 'x') {
                        uint64_t val = __builtin_va_arg(args, uint64_t);
                        print_num(val, 16);
                    }
                    break;
                }
                case 's': {
                    char *s = __builtin_va_arg(args, char*);
                    serial_puts(s);
                    break;
                }
                case '%':
                    serial_putc('%');
                    break;
            }
        } else {
            if (*fmt == '\n') serial_putc('\r');
            serial_putc(*fmt);
        }
        fmt++;
    }

    __builtin_va_end(args);
}

File: zig/src/serial.zig
// Serial port driver for COM1
const std = @import("std");

const COM1_PORT: u16 = 0x3F8;

// x86 I/O port functions
inline fn outb(port: u16, value: u8) void {
    asm volatile ("outb %[value], %[port]"
        :
        : [value] "{al}" (value),
          [port] "N{dx}" (port),
    );
}

inline fn inb(port: u16) u8 {
    return asm volatile ("inb %[port], %[result]"
        : [result] "={al}" (-> u8),
        : [port] "N{dx}" (port),
    );
}

pub fn init() void {
    outb(COM1_PORT + 1, 0x00); // Disable interrupts
    outb(COM1_PORT + 3, 0x80); // Enable DLAB
    outb(COM1_PORT + 0, 0x03); // Divisor low (38400 baud)
    outb(COM1_PORT + 1, 0x00); // Divisor high
    outb(COM1_PORT + 3, 0x03); // 8N1
    outb(COM1_PORT + 2, 0xC7); // Enable FIFO
    outb(COM1_PORT + 4, 0x0B); // IRQs enabled, RTS/DSR set
}

fn is_transmit_empty() bool {
    return (inb(COM1_PORT + 5) & 0x20) != 0;
}

pub fn write_byte(byte: u8) void {
    while (!is_transmit_empty()) {}
    outb(COM1_PORT, byte);
}

pub fn write_string(string: []const u8) void {
    for (string) |byte| {
        if (byte == '\n') {
            write_byte('\r');
        }
        write_byte(byte);
    }
}

// Hex printing helpers
const hex_digits = "0123456789ABCDEF";

pub fn write_hex_u64(value: u64) void {
    write_string("0x");
    var i: u6 = 60;
    while (true) : (i -%= 4) {
        const nibble = @as(u8, @truncate((value >> i) & 0xF));
        write_byte(hex_digits[nibble]);
        if (i == 0) break;
    }
}

pub fn write_hex_u32(value: u32) void {
    write_string("0x");
    var i: u5 = 28;
    while (true) : (i -%= 4) {
        const nibble = @as(u8, @truncate((value >> i) & 0xF));
        write_byte(hex_digits[nibble]);
        if (i == 0) break;
    }
}

pub fn write_dec_u64(value: u64) void {
    if (value == 0) {
        write_byte('0');
        return;
    }

    // Initialize buffer to avoid bare-metal stack issues
    var buf: [20]u8 = [_]u8{0} ** 20;
    var i: usize = buf.len;
    var n = value;

    while (n > 0) {
        i -= 1;
        buf[i] = @as(u8, @truncate(n % 10)) + '0';
        n /= 10;
    }

    write_string(buf[i..]);
}

pub fn write_dec_u32(value: u32) void {
    write_dec_u64(value);
}
